indexing-hadoop/src/main/java/io/druid/indexer/JobHelper.java

```diff
@@ -109,34 +121,96 @@ public static void setupClasspath(
     if (fs instanceof LocalFileSystem) {
       return;
     }
-
+    // Create distributed directory if it does not exist.
+    fs.mkdirs(distributedClassPath);
     for (String jarFilePath : jarFiles) {
       File jarFile = new File(jarFilePath);
       if (jarFile.getName().endsWith(".jar")) {
-        final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
-
-        if (!existing.contains(hdfsPath)) {
-          if (jarFile.getName().matches(".*SNAPSHOT(-selfcontained)?\\.jar$") || !fs.exists(hdfsPath)) {
-            log.info("Uploading jar to path[%s]", hdfsPath);
-            ByteStreams.copy(
-                Files.newInputStreamSupplier(jarFile),
-                new OutputSupplier<OutputStream>()
-                {
-                  @Override
-                  public OutputStream getOutput() throws IOException
-                  {
-                    return fs.create(hdfsPath);
-                  }
-                }
-            );
-          }
-
-          existing.add(hdfsPath);
+        if (isSnapshot(jarFile)) {
+          addSnapshotJarToClassPath(jarFile, intermediateClassPath, fs, conf);
+        } else {
+          addJarToClassPath(jarFile, distributedClassPath, intermediateClassPath, fs, conf);
         }
+      }
+    }
+  }
 
-        DistributedCache.addFileToClassPath(hdfsPath, conf, fs);
+  private static void addJarToClassPath(
+      File jarFile,
+      Path distributedClassPath,
+      Path intermediateClassPath,
+      FileSystem fs,
+      Configuration conf
+  )
+      throws IOException
+  {
+    // Non-snapshot jar files are uploaded to the shared classpath.
+    final Path hdfsPath = new Path(distributedClassPath, jarFile.getName());
+    if (!fs.exists(hdfsPath)) {
+      // Muliple jobs can try to upload the jar here,
+      // to avoid them from overwriting files, first upload to intermediateClassPath and then rename to the distributedClasspath.
+      final Path intermediateHdfsPath = new Path(intermediateClassPath, jarFile.getName());
+      uploadJar(jarFile, intermediateHdfsPath, fs);
+      try {
+        log.info("Renaming jar to path[%s]", hdfsPath);
+        fs.rename(intermediateHdfsPath, hdfsPath);
+        if(!fs.exists(hdfsPath)){
+          throw new ISE("File does not exist even after moving from[%s] to [%s]", intermediateHdfsPath, hdfsPath);
+        }
+      }
+      catch (IOException e) {
+        // rename failed, possibly due to race condition. check if some other job has uploaded the jar file.
+        if (!fs.exists(hdfsPath)) {
+          log.error(e, "IOException while Renaming jar file");
+          throw e;
+        }
+      } finally {
+        if (fs.exists(intermediateHdfsPath)) {
```

Comment:
ideally the exceptions would bubble up as suppressed exceptions, but that may not be possible to do in a clean way.